src/bundle.c

```diff
@@ -171,30 +175,242 @@ static int unload_tracked_bundle(const char *bundle_name, struct list **subs)
 	return EBUNDLE_NOT_TRACKED;
 }
 
-/* Check if bundle is included by any subscribed bundles */
-static bool is_included(const char *bundle_name, struct manifest *mom)
+/* Return list of bundles that include bundle_name */
+void required_by(struct list **reqd_by, const char *bundle_name, struct manifest *mom, int rl)
 {
-	bool ret = false;
 	struct list *b, *i;
+	// track recursion level for indentation
+	rl++;
 
 	b = list_head(mom->submanifests);
 	while (b) {
 		struct manifest *bundle = b->data;
 		b = b->next;
 
+		int indent = 0;
 		i = list_head(bundle->includes);
 		while (i) {
 			char *name = i->data;
 			i = i->next;
 
 			if (strcmp(name, bundle_name) == 0) {
-				ret = true;
-				goto out;
+				char *bundle_str = NULL;
+				indent = (rl - 1) * 4;
+				if (rl == 1) {
+					string_or_die(&bundle_str, "%*s* %s\n", indent + 2, "", bundle->component);
+				} else {
+					string_or_die(&bundle_str, "%*s|-- %s\n", indent, "", bundle->component);
+				}
+
+				*reqd_by = list_append_data(*reqd_by, bundle_str);
+				required_by(reqd_by, bundle->component, mom, rl);
 			}
 		}
 	}
+}
+/* Return recursive list of included bundles */
+int show_included_bundles(char *bundle_name)
+{
+	int ret;
+	int lock_fd;
+	int current_version = CURRENT_OS_VERSION;
+	struct list *subs = NULL;
+	struct list *deps = NULL;
+	struct manifest *mom = NULL;
+
+	ret = swupd_init(&lock_fd);
+	if (ret != 0) {
+		fprintf(stderr, "Failed updater initialization, exiting now.\n");
+		return ret;
+	}
+
+	current_version = get_current_version(path_prefix);
+	if (current_version < 0) {
+		fprintf(stderr, "Error: Unable to determine current OS version\n");
+		ret = ECURRENT_VERSION;
+		goto out;
+	}
+
+	mom = load_mom(current_version, false);
+	if (!mom) {
+		fprintf(stderr, "Cannot load official manifest MoM for version %i\n", current_version);
+		ret = EMOM_NOTFOUND;
+		goto out;
+	}
+
+	// add_subscriptions takes a list, so construct one with only bundle_name
+	struct list *bundles = NULL;
+	bundles = list_prepend_data(bundles, bundle_name);
+	ret = add_subscriptions(bundles, &subs, current_version, mom, true, 0);
+	list_free_list(bundles);
+	if (ret != add_sub_NEW) {
+		// something went wrong or there were no includes, print a message and exit
+		const char *m;
+		if (ret & add_sub_ERR) {
+			m = "Processing error";
```

Comment:
... it is assigned to. I don't recall the C language semantics for this one, but I would think assigning to a `const char *` in a later statement would be forbidden. Testing this with an int type (`const int a; a = 1;`) results in a compiler error.

If someone can chime in as to why the const is apparently ignored in this case, please do so.

Regardless, it's more common in the swupd code base to dynamically allocate strings with `string_or_die()`, so I would take that route.